File: demo-python-sync/demo_private.py
# -*- coding: utf-8 -*-
import json
import time
import requests
import sys

try:
    from urllib.parse import urlparse
except:
    # noinspection PyUnresolvedReferences
    from urlparse import urlparse
import hmac
import hashlib


class Secret:
    ot_key = ''
    ot_secret = ''


def gen_nonce():
    return str(int(time.time() * 1000000))


py3 = sys.version_info > (3, 0)


def gen_sign(secret, verb, endpoint, nonce, data_str):
    # Parse the url so we can remove the base and extract just the path.

    if data_str is None:
        data_str = ''

    parsed_url = urlparse(endpoint)
    path = parsed_url.path

    # print "Computing HMAC: %s" % verb + path + str(nonce) + data
    message = verb + path + str(nonce) + data_str

    if py3:
        signature = hmac.new(bytes(secret, 'utf8'), bytes(message, 'utf8'), digestmod=hashlib.sha256).hexdigest()
    else:
        signature = hmac.new(secret, message, hashlib.sha256).hexdigest()
    return signature


def api_call(method, endpoint, params=None, data=None, timeout=15, host='https://1token.trade/api/v1/trade'):
    assert params is None or isinstance(params, dict)
    assert data is None or isinstance(data, dict)
    method = method.upper()

    nonce = gen_nonce()

    url = host + endpoint

    json_str = json.dumps(data) if data else ''
    sign = gen_sign(Secret.ot_secret, method, endpoint, nonce, json_str)
    headers = {'Api-Nonce': str(nonce), 'Api-Key': Secret.ot_key, 'Api-Signature': sign,
               'Content-Type': 'application/json'}
    res = requests.request(method, url=url, data=json_str, params=params, headers=headers, timeout=timeout)
    return res


def demo(account):
    print('查看账户信息')
    r = api_call('GET', '/{}/info'.format(account))
    print(r.json())

    print('撤销所有订单')
    r = api_call('DELETE', '/{}/orders/all'.format(account))
    print(r.json())

    print('下单')
    r = api_call('POST', '/{}/orders'.format(account),
                 data={'contract': 'okex/btc.usdt', 'price': 10, 'bs': 'b', 'amount': 1, 'options': {'close': True}})
    print(r.json())
    assert r.json()['client_oid']
    exg_oid = r.json()['exchange_oid']

    print('查询挂单 应该有一个挂单')
    r = api_call('GET', '/{}/orders'.format(account))
    print(r.json())
    assert len(r.json()) == 1

    print('用 exchange oid撤单')
    r = api_call('DELETE', '/{}/orders'.format(account), params={'exchange_oid': exg_oid})
    print(r.json())

    print('查询挂单 应该没有挂单')
    r = api_call('GET', '/{}/orders'.format(account))
    print(r.json())
    assert len(r.json()) == 0


def main():
    ot_key = input('ot-key: ')
    ot_secret = input('ot-secret: ')
    account = input('请输入交易账号 账号格式是 {交易所}/{交易账户名} 比如 okex/mock-1token: ')
    Secret.ot_key = ot_key
    Secret.ot_secret = ot_secret

    demo(account)


if __name__ == '__main__':
    main()

File: demo-node-sync/demo_private.js
const request = require('sync-request')
const crypto=require('crypto')
const url=require('url')
const assert = require('assert')
const readline = require('readline').createInterface({
     input: process.stdin,
     output: process.stdout
   })
  
  
const secret={otKey:'',otSecret:''}
function getHash(secret,message){
	console.log('hash',crypto.createHmac('SHA256', secret).update(message).digest('hex'))
    return crypto.createHmac('SHA256', secret).update(message).digest('hex')
}
function genSign(secret, verb, endpoint, nonce, dataStr){
    if(!dataStr) dataStr=''
    let parsedUrl = url.parse(endpoint)
    path = parsedUrl.path
    message = verb + path + nonce + dataStr
	console.log(message)
    return getHash(secret,message)
}
function apiCall(method, endpoint, params={}, data=null, timeout=15000, host='https://1token.trade/api/v1/trade'){
    method = method.toUpperCase()

    let nonce = genNonce()

    let url = host + endpoint

    let jsonStr = data?JSON.stringify(data):''
    sign = genSign(secret.otSecret, method, endpoint, nonce, jsonStr)
	console.log(secret,method,endpoint,nonce,jsonStr,sign)
    headers = {'Api-Nonce': nonce+'', 'Api-Key': secret.otKey, 'Api-Signature': sign,
               'Content-Type': 'application/json'}
    let options={body:jsonStr,qs:params,headers,timeout}
    res = request(method, url=url, options)
    return res
}
function genNonce(){
    return Date.now() * 1000
}

function demo(account){
    console.log('查看账户信息')
    r = apiCall('GET', `/${account}/info`)
    console.log(r.getBody('utf8'))

    console.log('撤销所有订单')
    r = apiCall('DELETE',`/${account}/orders/all`)
    console.log(r.getBody('utf8'))

    console.log('下单')
    r = apiCall('POST',`/${account}/orders`,{},
                 data={'contract': 'okex/btc.usdt', 'price': 10, 'bs': 'b', 'amount': 1, 'options': {'close': true}})
    console.log(r.getBody('utf8'))
    r=JSON.parse(r.getBody('utf8'))
    assert(r.client_oid)
    let exgOid=r.exchange_oid
	console.log('查询挂单 应该有一个挂单')
	r = apiCall('GET', `/${account}/orders`)
	console.log(r.getBody('utf8'))
	r=JSON.parse(r.getBody('utf8'))
	assert(r.length===1)
	console.log('用 exchange oid撤单')
	r = apiCall('DELETE', `/${account}/orders`, params={exchange_oid: exgOid})
	console.log(r.getBody('utf8'))
	console.log('查询挂单 应该没有挂单')
	r = apiCall('GET', `/${account}/orders`)
	console.log(r.getBody('utf8'))
	r=JSON.parse(r.getBody('utf8'))
	assert(r.length===0)
    

}
function main(){
    readline.question('ot-key: ', otKey => {
        secret.otKey=otKey
        readline.question('ot-secret: ',otSecret=>{
            secret.otSecret=otSecret
            readline.question('请输入交易账号 账号格式是 {交易所}/{交易账户名} 比如 okex/mock-1token: ',account=>{
                readline.close()
                demo(account)
            })
        })
    })
}
main()
